File: apps/workers/shared/src/api-key-generator.ts
/**
 * Extract the key prefix (first 12 characters)
 * @param fullKey - Full API key
 * @returns Key prefix (e.g., "requiem_abc1")
 */
export function extractKeyPrefix(fullKey: string): string {
  return fullKey.substring(0, 12);
}

/**
 * Validate key format
 * @param key - Key to validate
 * @returns true if valid format
 */
export function isValidKeyFormat(key: string): boolean {
  // Must be requiem_ followed by 24 characters
  const pattern = /^requiem_[0-9a-zA-Z]{24}$/;
  return pattern.test(key);
}

File: apps/dashboard/app/services/api_key_generator.rb
# frozen_string_literal: true

require "bcrypt"

class ApiKeyGenerator
  # Generate a new API key in the format: rq_live_<24_random_chars>
  # Returns the full key (which should be shown to the user once)
  def self.generate(environment: :live)
    prefix = environment == :test ? "rq_test" : "rq_live"
    random_part = Nanoid.generate(size: 24, alphabet: "0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ")

    "#{prefix}_#{random_part}"
  end

  # Extract the prefix for display (first 12 characters)
  def self.extract_prefix(full_key)
    full_key[0..11] if full_key
  end

  # Hash the full key for secure storage
  def self.hash_key(full_key)
    BCrypt::Password.create(full_key)
  end

  # Verify a key against a hash
  def self.verify_key(full_key, key_hash)
    BCrypt::Password.new(key_hash) == full_key
  rescue BCrypt::Errors::InvalidHash
    false
  end
end
